### PR TITLE
Alerts management new alert processing rule

### PR DESCRIPTION
**Pull request by Liran-E**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required) - pylint never returns
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? where is it?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Alerts management

File: src/alertsmanagement/setup.cfg [hunk 1]
```diff
@@ -1,2 +1 @@
-[bdist_wheel]
-universal=1
+#setup.cfg
```
Comment (wangzelin007) [inline src/alertsmanagement/setup.cfg:1]: Please add new line at the end of file.

File: src/alertsmanagement/azext_alertsmanagement/vendored_sdks/alertsmanagement/py.typed [hunk 2]
```diff
@@ -0,0 +1 @@
+# Marker file for PEP 561.
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/vendored_sdks/alertsmanagement/py.typed:1]: Please add new line at the end of file.

File: src/alertsmanagement/azext_alertsmanagement/azext_metadata.json [hunk 3]
```diff
@@ -1,4 +1,4 @@
 {
-    "azext.isExperimental": true,
-    "azext.minCliCoreVersion": "2.3.1"
+    "azext.isPreview": true,
+    "azext.minCliCoreVersion": "2.15.0"
 }
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/azext_metadata.json:4]: Please add new line at the end of file.

File: src/alertsmanagement/azext_alertsmanagement/_help.py [hunk 4]
```diff
@@ -1,72 +1,191 @@
...
     type: group
-    short-summary: Commands to manage action rule.
+    short-summary: Manage alert processing rule with alertsmanagement
 """
 
-helps['monitor action-rule create'] = """
+helps['monitor alert-processing-rule delete'] = """
     type: command
-    short-summary: Create an action rule
+    short-summary: "Delete an alert processing rule."
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_help.py:23]: May I ask why do you use double quotes in short-summary ?

File: src/alertsmanagement/azext_alertsmanagement/commands.py [hunk 5]
```diff
@@ -12,17 +12,16 @@
...
+    alertsmanagement_processing_rules = CliCommandType(
+        operations_tmpl='azext_alertsmanagement.vendored_sdks.alertsmanagement.operations._alert_processing_rules_operations'
+                        '#AlertProcessingRulesOperations.{}',
+        client_factory=cf_processing_rules)
+    with self.command_group('monitor alert-processing-rule', alertsmanagement_processing_rules, client_factory=cf_processing_rules,
+                            is_preview=True) as g:
+        g.custom_command('create', 'create_alertsmanagement_processing_rule')
+        g.generic_update_command('update', custom_func_name='update_alertsmanagement_processing_rule',
+                                setter_arg_name='alert_processing_rule', getter_name='get_by_name')
+        g.custom_command('delete', 'delete_alertsmanagement_processing_rule')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/commands.py:25]: Do we need to consider adding confirmation=True for the delete operation ?

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 6]
```diff
@@ -1,4 +1,84 @@
...
+
+def validate_datetime_format(namespace):
+    format = '%Y-%m-%d %H:%M:%S' #'%m/%d/%Y'
+    if namespace.schedule_start_datetime:
+        datetime.strptime(namespace.schedule_start_datetime, format)
+    if namespace.schedule_end_datetime:
+        datetime.strptime(namespace.schedule_end_datetime, format)
+        if namespace.schedule_start_datetime:
+            if namespace.schedule_end_datetime < namespace.schedule_start_datetime:
+                raise CLIError('Argument Error: end-date is before start-date')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:12]:
```suggestion
                raise ValidationError('Argument Error: end-date is before start-date')
```

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 7]
```diff
@@ -1,4 +1,84 @@
...
+        datetime.strptime(namespace.schedule_recurrence_start_time, format)
+    if namespace.schedule_recurrence_end_time:
+        datetime.strptime(namespace.schedule_recurrence_end_time, format)
+
+def validate_severity(namespace):
+    if namespace.filter_severity:
+        validate_only_equals_operator(namespace.filter_severity)
+        for x in namespace.filter_severity[1:]:
+            if x not in ['Sev0', 'Sev1', 'Sev2', 'Sev3', 'Sev4']:
+                raise CLIError('Argument Error: filter-severity values have to be one of [Equals, NotEquals, Sev0, Sev1, Sev2, Sev3, Sev4]')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:26]:
```suggestion
                raise InvalidArgumentValueError('Argument Error: filter-severity values have to be one of [Equals, NotEquals, Sev0, Sev1, Sev2, Sev3, Sev4]')
```

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 8]
```diff
@@ -1,4 +1,84 @@
...
+        for x in namespace.filter_monitor_condition[1:]:
+            if x not in ['Fired', 'Resolved']:
+                raise CLIError('Argument Error: filter-monitor-condition values have to be one of [Equals, NotEquals, Fired, Resolved]')
+
+def validate_signal_type(namespace):
+    if namespace.filter_signal_type:
+        validate_only_equals_operator(namespace.filter_signal_type)
+        for x in namespace.filter_signal_type[1:]:
+            if x not in ['Metric', 'Log', 'Unknown']:
+                raise CLIError('Argument Error: filter-signal-type values have to be one of [Equals, NotEquals, Metric, Log, Unknown]')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:40]:
```suggestion
                raise InvalidArgumentValueError('Argument Error: filter-signal-type values have to be one of [Equals, NotEquals, Metric, Log, Unknown]')
```

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 9]
```diff
@@ -1,4 +1,84 @@
...
+    if namespace.filter_resource_group:
+        validate_full_operator(namespace.filter_resource_group)
+
+def validate_resource_type(namespace):
+    if namespace.filter_resource_type:
+        validate_only_equals_operator(namespace.filter_resource_type)
+
+def validate_full_operator(args):
+    if len(args) < 2:
+        raise CLIError('Filter Argument Error: values length can\'t be smaller than 2')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:76]:
```suggestion
        raise RequiredArgumentMissingError('Filter Argument Error: values length can\'t be smaller than 2')
```

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 10]
```diff
@@ -1,4 +1,84 @@
...
+    if args[0].lower() not in ['equals', 'notequals', 'contains', 'doesnotcontain']:
+        raise CLIError('Filter Argument Error: operator must be one of the follows: Equals, NotEquals, Contains, DoesNotContain')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:78]:
```suggestion
        raise InvalidArgumentValueError('Filter Argument Error: operator must be one of the follows: Equals, NotEquals, Contains, DoesNotContain')
```

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 11]
```diff
@@ -1,4 +1,84 @@
...
+
+def validate_only_equals_operator(args):
+    if len(args) < 2:
+        raise CLIError('Filter Argument Error: values length can\'t be smaller than 2')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:82]:
```suggestion
        raise RequiredArgumentMissingError('Filter Argument Error: values length can\'t be smaller than 2')
```

File: src/alertsmanagement/azext_alertsmanagement/_validators.py [hunk 12]
```diff
@@ -1,4 +1,84 @@
...
+    if args[0].lower() not in ['equals', 'notequals']:
+        raise CLIError('Filter Argument Error: operator must be one of the follows: Equals, NotEquals')
```
Comment (wangzelin007) [inline src/alertsmanagement/azext_alertsmanagement/_validators.py:84]:
```suggestion
        raise InvalidArgumentValueError('Filter Argument Error: operator must be one of the follows: Equals, NotEquals')
```

File: src/alertsmanagement/azext_alertsmanagement/_help.py (hunk 4 above)
Comment (Liran-E) [reply src/alertsmanagement/azext_alertsmanagement/_help.py:18]: I removed all double quotes from short-summary of commands.
Should I do so for short-summary of parameters?

File: src/alertsmanagement/azext_alertsmanagement/commands.py (hunk 5 above)
Comment (Liran-E) [reply src/alertsmanagement/azext_alertsmanagement/commands.py]: Thanks for the suggestion. Added confirmation to the delete command

File: src/alertsmanagement/azext_alertsmanagement/action.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+# pylint: disable=unused-wildcard-import
+
+from .generated.action import *  # noqa: F403
+try:
+    from .manual.action import *  # noqa: F403
+except ImportError as e:
+    if e.name.endswith('manual.action'):
+        pass
+    else:
+        raise e
```
Comment (kairu-ms) [inline src/alertsmanagement/azext_alertsmanagement/action.py:20]: You can delete this file. It's not needed.

File: src/alertsmanagement/setup.py [hunk 14]
```diff
@@ -8,15 +8,13 @@
...
-    logger.warn("Wheel is not available, disabling bdist_wheel hook")
 
-# TODO: Confirm this is the right version number you want and it matches your
 # HISTORY.rst entry.
-VERSION = '0.1.1'
+VERSION = '0.2.0'
+try:
+    from azext_alertsmanagement.manual.version import VERSION
+except ImportError:
+    pass
```
Comment (kairu-ms) [inline src/alertsmanagement/setup.py:17]: You can remove this line
```suggestion
```

File: src/alertsmanagement/azext_alertsmanagement/manual/__init__.py [hunk 15]
```diff
@@ -9,5 +9,4 @@
 # regenerated.
 # --------------------------------------------------------------------------
 
-VERSION = "0.2.0rc2"
-
+__path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
Comment (kairu-ms) [inline src/alertsmanagement/azext_alertsmanagement/manual/__init__.py:12]: You can remove this file.

File: src/alertsmanagement/setup.py [hunk 16]
```diff
@@ -8,15 +8,9 @@
 
 from codecs import open
 from setuptools import setup, find_packages
-try:
-    from azure_bdist_wheel import cmdclass
-except ImportError:
-    from distutils import log as logger
-    logger.warn("Wheel is not available, disabling bdist_wheel hook")
```
Comment (kairu-ms) [inline src/alertsmanagement/setup.py:11]:
```suggestion
try:
    from azure_bdist_wheel import cmdclass
except ImportError:
    from distutils import log as logger
    logger.warn("Wheel is not available, disabling bdist_wheel hook")

```

File: src/alertsmanagement/azext_alertsmanagement/__init__.py [hunk 17]
```diff
@@ -1,32 +1,53 @@
...
         from azext_alertsmanagement.commands import load_command_table
         load_command_table(self, args)
+        try:
+            from azext_alertsmanagement.manual.commands import load_command_table as load_command_table_manual
+            load_command_table_manual(self, args)
+        except ImportError as e:
+            if e.name.endswith('manual.commands'):
+                pass
+            else:
+                raise e
```
Comment (kairu-ms) [inline src/alertsmanagement/azext_alertsmanagement/__init__.py:37]:
```suggestion
```

File: src/alertsmanagement/azext_alertsmanagement/__init__.py [hunk 18]
```diff
@@ -1,32 +1,53 @@
...
         from azext_alertsmanagement._params import load_arguments
         load_arguments(self, command)
+        try:
+            from azext_alertsmanagement.manual._params import load_arguments as load_arguments_manual
+            load_arguments_manual(self, command)
+        except ImportError as e:
+            if e.name.endswith('manual._params'):
+                pass
+            else:
+                raise e
```
Comment (kairu-ms) [inline src/alertsmanagement/azext_alertsmanagement/__init__.py:42]:
```suggestion
```

File: src/alertsmanagement/azext_alertsmanagement/__init__.py [hunk 19]
```diff
@@ -1,23 +1,28 @@
...
-        from azext_alertsmanagement._client_factory import cf_alertsmanagement
+        from azext_alertsmanagement._client_factory import cf_alertsmanagement_cl
         alertsmanagement_custom = CliCommandType(
             operations_tmpl='azext_alertsmanagement.custom#{}',
-            client_factory=cf_alertsmanagement)
-        super(AlertsCommandsLoader, self).__init__(cli_ctx=cli_ctx,
-                                                   custom_command_type=alertsmanagement_custom)
+            client_factory=cf_alertsmanagement_cl)
+        parent = super(AlertsManagementClientCommandsLoader, self)
+        parent.__init__(cli_ctx=cli_ctx, custom_command_type=alertsmanagement_custom)
```
Comment (kairu-ms) [inline src/alertsmanagement/azext_alertsmanagement/__init__.py:25]:
```suggestion
        super().__init__(cli_ctx=cli_ctx, custom_command_type=alertsmanagement_custom)
```

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**